File: 05-failure/02-timeout-budgets-and-deadline-propagation/python/deadline_chain.py
"""
Layer 5 - Topic 2: deadline propagation through a three-hop chain, in one
Python process.

Python is the language where this topic has a second half nobody expects.
Cancelling a coroutine cancels it at its next `await` -- it does NOT stop
work that has already been handed to something else. The query you already
sent keeps running on the database server until it finishes or
`statement_timeout` kills it, holding its pooled connection the whole time.
So a Python deadline discipline has two halves, and variant 3 below has
only the first one.

WHAT THIS DEMONSTRATES
  gateway -> service_b -> service_c, where C holds a pooled connection for
  a controlled service time. The gateway's budget is 500ms. Four variants:

    1. naive, C healthy       everything succeeds; the bug is invisible
    2. naive, C slow          one query in four takes 800ms; the gateway
                              504s at 500ms and C works on anyway
    3. deadline propagated    B and C refuse work that cannot finish, and
                              hand back a connection the moment they find
                              the request behind it is already dead
    4. + statement_timeout    the query itself is bounded, not just the
                              coroutine awaiting it

WHAT TO LOOK FOR IN THE OUTPUT
  1. `zombie/s` is zero in variant 1, large in variant 2. A zombie is a
     completion C finished AFTER the gateway had already returned 504:
     one pool slot, one full service time, zero value.
  2. `C pool in use` is pinned at the pool size in variant 2. That is
     topic 1's L, consumed entirely by work nobody is waiting for.
  3. Variant 3 helps and does not fix it. This is the Python-specific
     finding, and it is the point of the whole file: the ContextVar
     stopped your Python, and the 800ms query kept running to the end
     holding its connection. You shed load in the app and left the
     database pinned -- the worst of both, since you get the errors AND
     keep the load.
  4. Variant 4 is the one that frees the pool, because it bounds the work
     at the resource rather than at the caller. Compare the gateway
     success column across 2, 3 and 4 -- that is the number that pays for
     all of this.

The load generator is OPEN MODEL. It does not wait for a response before
sending the next request, because a real client does not either, and a
closed-loop generator would quietly hide every effect below.

RUN
    python3 deadline_chain.py
"""
from __future__ import annotations

import asyncio
import contextvars
import statistics
import time

# ---------------------------------------------------------------- config

GATEWAY_BUDGET = 0.500      # what the gateway promises its own caller
SLACK = 0.020               # subtracted per hop; also the reject-now floor
HOP_OVERHEAD = 0.005        # B's and C's own work, before the next hop
C_SERVICE_FAST = 0.040      # the ordinary query
C_SERVICE_SLOW = 0.800      # the same query when the dependency is unwell
SLOW_FRACTION = 0.25        # a slow dependency is usually slow for a SUBSET
C_POOL_SIZE = 8
RATE = 50.0                 # offered requests per second, Poisson arrivals
DURATION = 12.0
GAUGE_EVERY = 0.020

# Absolute deadline, in perf_counter time, for the request being handled on
# this task. contextvars is the only ambient carrier Python has, and an
# ABSOLUTE deadline is the load-bearing choice: a per-call `timeout=`
# argument cannot compose, because the third call in a handler has no idea
# what the first two already spent.
DEADLINE: contextvars.ContextVar[float | None] = contextvars.ContextVar(
    "deadline", default=None
)


def remaining() -> float:
    """Budget left for the current request, or infinity if nobody said."""
    d = DEADLINE.get()
    return float("inf") if d is None else d - time.perf_counter()


# ------------------------------------------------------------ the metrics


class Metrics:
    def __init__(self) -> None:
        self.gateway_ok = 0
        self.gateway_timeout = 0
        self.gateway_rejected = 0     # refused up front for lack of budget
        self.c_started = 0
        self.c_completed = 0
        self.c_zombie = 0             # completed after its gateway gave up
        self.c_killed = 0             # statement_timeout got there first
        self.c_abandoned = 0          # got a connection, gave it straight back
        self.c_latencies: list[float] = []
        self.pool_gauge: list[int] = []

    def row(self, label: str, seconds: float, pool_size: int) -> str:
        total = self.gateway_ok + self.gateway_timeout + self.gateway_rejected
        success = 100.0 * self.gateway_ok / total if total else 0.0
        p99 = _percentile(sorted(self.c_latencies), 99) * 1000
        gauge = statistics.fmean(self.pool_gauge) if self.pool_gauge else 0.0
        return (
            f"{label:<28} {success:>9.1f}% {self.c_zombie / seconds:>9.1f} "
            f"{gauge:>8.1f}/{pool_size:<5} {p99:>9.0f} "
            f"{self.c_killed / seconds:>9.1f} {self.c_abandoned / seconds:>11.1f}"
        )


def _percentile(sorted_values: list[float], p: float) -> float:
    if not sorted_values:
        return 0.0
    k = round(p / 100 * (len(sorted_values) - 1))
    return sorted_values[min(k, len(sorted_values) - 1)]


# ------------------------------------------------- hop 3, and its database


class Database:
    """
    A pool of connections and a server that does not care about your
    coroutines. `query` holds a connection for the whole of its duration.
    Cancelling the *caller* does not shorten it -- only `statement_timeout`
    does, and only when one was set.
    """

    def __init__(self, size: int, metrics: Metrics) -> None:
        self.sem = asyncio.Semaphore(size)
        self.in_use = 0
        self.metrics = metrics

    async def query(self, duration: float, deadline: float | None,
                    use_statement_timeout: bool) -> bool:
        """Returns True if the query ran to completion, False if it did not."""
        async with self.sem:
            # Checked out. Everything below happens INSIDE the transaction,
            # which is the only place `SET LOCAL` means anything.
            now = time.perf_counter()
            if deadline is not None and deadline - now < SLACK:
                # The request that queued for this connection died while it
                # was queueing. Give the connection straight back rather
                # than spending a service time on a corpse. Under overload
                # this is where most of the recovered capacity comes from.
                self.metrics.c_abandoned += 1
                return False

            self.in_use += 1
            try:
                stmt = None
                if deadline is not None and use_statement_timeout:
                    # Derived from the SAME number as the application budget.
                    # Two independently chosen timeouts is how you get a
                    # service that sheds load while Postgres stays pinned.
                    stmt = max(0.0, deadline - now - SLACK)

                if stmt is None or stmt >= duration:
                    await asyncio.sleep(duration)
                    return True
                await asyncio.sleep(stmt)
                self.metrics.c_killed += 1
                return False
            finally:
                self.in_use -= 1


async def service_c(db: Database, m: Metrics, slow: bool,
                    propagate: bool, use_statement_timeout: bool,
                    gateway_deadline: float) -> None:
    if propagate and remaining() < SLACK:
        # Refuse to START work that cannot finish. This is the cheapest win
        # in the whole layer: a request rejected here costs no pool slot,
        # no queue position, nothing at all.
        raise TimeoutError("no budget left at C")

    await asyncio.sleep(HOP_OVERHEAD)

    duration = C_SERVICE_SLOW if slow else C_SERVICE_FAST
    started = time.perf_counter()
    m.c_started += 1

    # THE PYTHON-SPECIFIC PART. The query runs as its own task. When the
    # caller's `wait_for` fires, THIS coroutine is cancelled -- and the task
    # below is not, because it is shielded. That is precisely what happens
    # with a real driver: your `await` is abandoned, the statement on the
    # server is not.
    task = asyncio.create_task(
        db.query(duration, DEADLINE.get() if propagate else None,
                 use_statement_timeout)
    )

    def account(_fut: asyncio.Future) -> None:
        done = time.perf_counter()
        m.c_completed += 1
        m.c_latencies.append(done - started)
        if done > gateway_deadline:
            m.c_zombie += 1

    task.add_done_callback(account)
    await asyncio.shield(task)


# ------------------------------------------------------------ hops 2 and 1


async def service_b(db: Database, m: Metrics, slow: bool,
                    propagate: bool, use_statement_timeout: bool,
                    gateway_deadline: float) -> None:
    if propagate:
        left = remaining()
        if left < SLACK:
            raise TimeoutError("no budget left at B")

    await asyncio.sleep(HOP_OVERHEAD)

    if propagate:
        # budget_out = budget_in - elapsed_here - slack
        out = max(0.0, remaining() - SLACK)
    else:
        # The bug, and it looks completely reasonable on the page: a
        # constant, the same one the gateway used, chosen once and copied.
        out = GATEWAY_BUDGET

    await asyncio.wait_for(
        service_c(db, m, slow, propagate, use_statement_timeout,
                  gateway_deadline),
        timeout=out,
    )


async def gateway(db: Database, m: Metrics, slow: bool,
                  propagate: bool, use_statement_timeout: bool) -> None:
    deadline = time.perf_counter() + GATEWAY_BUDGET
    if propagate:
        # The one line that starts the whole discipline. Downstream reads it
        # from the context; over a real network it becomes a header.
        DEADLINE.set(deadline)

    try:
        await asyncio.wait_for(
            service_b(db, m, slow, propagate, use_statement_timeout, deadline),
            timeout=GATEWAY_BUDGET,
        )
        m.gateway_ok += 1
    except TimeoutError:
        # asyncio.wait_for raises TimeoutError on expiry, and our own
        # "refuse to start" path raises the same type on purpose: from the
        # gateway's seat they are the same outcome with very different costs.
        m.gateway_timeout += 1


# ------------------------------------------------------------- the driver


async def run_variant(slow_fraction: float, propagate: bool,
                      use_statement_timeout: bool) -> Metrics:
    m = Metrics()
    db = Database(C_POOL_SIZE, m)
    loop = asyncio.get_running_loop()
    stop = asyncio.Event()

    async def sample_pool() -> None:
        while not stop.is_set():
            await asyncio.sleep(GAUGE_EVERY)
            m.pool_gauge.append(db.in_use)

    sampler = asyncio.create_task(sample_pool())

    import random
    rng = random.Random(20250502)   # same arrivals and same slow requests
    begin = loop.time()             # in every variant, so the comparison
    end = begin + DURATION          # is of the policy and nothing else
    at = begin
    tasks: list[asyncio.Task] = []

    while True:
        at += rng.expovariate(RATE)
        if at > end:
            break
        delay = at - loop.time()
        if delay > 0:
            await asyncio.sleep(delay)
        slow = rng.random() < slow_fraction
        # Each request runs in its own context, so the ContextVar set inside
        # the gateway belongs to that request and nothing else.
        tasks.append(
            asyncio.create_task(
                gateway(db, m, slow, propagate, use_statement_timeout),
                context=contextvars.copy_context(),
            )
        )

    await asyncio.gather(*tasks, return_exceptions=True)
    # Drain. Zombies are by definition still running after everyone gave up,
    # so a report taken at the end of the load would undercount them.
    await asyncio.sleep(C_SERVICE_SLOW + 0.3)
    stop.set()
    await sampler
    return m


HEADER = (f"{'variant':<28} {'gw success':>10} {'zombie/s':>9} "
          f"{'C pool in use':>14} {'C p99 ms':>9} {'killed/s':>9} {'gaveback/s':>11}")


async def main() -> None:
    fast_demand = RATE * (1 - SLOW_FRACTION) * C_SERVICE_FAST
    slow_demand = RATE * SLOW_FRACTION * C_SERVICE_SLOW
    print("Deadline propagation through gateway -> service_b -> service_c.")
    print(f"Gateway budget {GATEWAY_BUDGET*1000:.0f}ms, slack {SLACK*1000:.0f}ms/hop, "
          f"C pool {C_POOL_SIZE}, offered {RATE:.0f} rps for {DURATION:.0f}s.")
    print(f"When C is unwell, {SLOW_FRACTION:.0%} of queries take "
          f"{C_SERVICE_SLOW*1000:.0f}ms and the rest take {C_SERVICE_FAST*1000:.0f}ms.")
    print(f"Demand on the pool is then {slow_demand:.1f} + {fast_demand:.1f} = "
          f"{slow_demand + fast_demand:.1f} connection-seconds per second "
          f"against {C_POOL_SIZE} available,")
    print(f"i.e. rho = {(slow_demand + fast_demand) / C_POOL_SIZE:.2f}. The slow "
          f"queries alone are {slow_demand / C_POOL_SIZE:.0%} of the pool, and "
          f"none of them can beat the budget.\n")
    print(HEADER)
    print("-" * len(HEADER))

    variants = [
        ("1 naive, C healthy", 0.0, False, False),
        ("2 naive, C slow", SLOW_FRACTION, False, False),
        ("3 propagated", SLOW_FRACTION, True, False),
        ("4 propagated + stmt_timeout", SLOW_FRACTION, True, True),
    ]
    for label, frac, prop, stmt in variants:
        m = await run_variant(frac, prop, stmt)
        print(m.row(label, DURATION, C_POOL_SIZE))

    print()
    print("Rows 2 and 3: propagation stops C queueing work whose caller has")
    print("already gone, and hands connections back the moment it finds one")
    print("checked out for a dead request ('gaveback/s'). That is real, and on")
    print("its own it is not enough.")
    print()
    print("Rows 3 and 4 are the Python footnote, and the reason this file is")
    print("written in Python. Cancelling the coroutine stopped your code; the")
    print("800ms query kept running to the end, holding its connection, and")
    print("25% of arrivals doing that is more than the pool has to give. Only")
    print("row 4 bounds the work where the work actually is. Shedding in the")
    print("app while the database stays pinned is the worst of both outcomes:")
    print("you get the errors AND you keep the load.")
    print()
    print("Every zombie completion is one pool slot for one service time,")
    print("producing a response no client will read. That is topic 1's rho,")
    print("climbing for work worth nothing.")


if __name__ == "__main__":
    asyncio.run(main())

File: 05-failure/02-timeout-budgets-and-deadline-propagation/nodejs/deadline_chain.js
// Layer 5 - Topic 2: deadline propagation through a three-hop chain, in one
// Node process.
//
// Node is the runtime that comes closest to correct-by-default here, because
// the primitives are actually in the platform rather than in a library:
// AbortSignal.timeout(ms) for a local bound, AbortSignal.any([...]) to
// combine an inbound signal with it, and req.signal on the server so a
// disconnected client propagates without anyone opting in. Nothing forces
// you to thread the signal through, though, so the real failure mode is a
// forgotten `{ signal }` on one call out of nine -- and this file shows what
// that one forgotten argument costs.
//
// WHAT THIS DEMONSTRATES
//   gateway -> serviceB -> serviceC, where C holds a pooled connection for a
//   controlled service time. The gateway's budget is 500ms. Four variants:
//
//     1. naive, C healthy       everything succeeds; the bug is invisible
//     2. naive, C slow          one query in four takes 800ms; the gateway
//                               aborts at 500ms and C works on anyway
//     3. deadline propagated    B and C refuse work that cannot finish and
//                               hand a connection straight back when the
//                               request behind it is already dead
//     4. + statement timeout    the query itself is bounded, not just the
//                               promise awaiting it
//
// WHAT TO LOOK FOR IN THE OUTPUT
//   1. `zombie/s` is zero in variant 1 and large in variant 2. A zombie is
//      a completion C finished AFTER the gateway had already given up: one
//      pool slot, one full service time, zero value.
//   2. `C pool in use` is pinned at the pool size in variant 2. That is
//      topic 1's L, spent entirely on work nobody is waiting for.
//   3. Variant 3 helps and does not fix it, and the reason is the same in
//      Node as it is in Python: aborting a signal aborts YOUR await. The
//      work already handed to something else -- a query on a database
//      server, a job on the libuv thread pool -- runs to completion.
//   4. Variant 4 bounds the work at the resource. Watch gateway success go
//      up in the same row that `C pool in use` comes down; those are the
//      same fact stated twice.
//
// The load generator is OPEN MODEL: arrivals are a Poisson process and the
// generator does not wait for a response before sending the next request.
//
// RUN
//   node deadline_chain.js

'use strict';

// ------------------------------------------------------------------ config

const GATEWAY_BUDGET_MS = 500;   // what the gateway promises its own caller
const SLACK_MS = 20;             // subtracted per hop; also the reject-now floor
const HOP_OVERHEAD_MS = 5;       // B's and C's own work, before the next hop
const C_SERVICE_FAST_MS = 40;    // the ordinary query
const C_SERVICE_SLOW_MS = 800;   // the same query when the dependency is unwell
const SLOW_FRACTION = 0.25;      // a slow dependency is usually slow for a SUBSET
const C_POOL_SIZE = 8;
const RATE = 50;                 // offered requests per second
const DURATION_MS = 12000;
const GAUGE_EVERY_MS = 20;

const sleep = (ms) => new Promise((r) => setTimeout(r, Math.max(0, ms)));

// A tiny deterministic PRNG so every variant sees the identical arrival
// pattern and the identical set of slow requests. What differs between the
// rows below is policy, and only policy.
function mulberry32(seed) {
  return function () {
    seed |= 0;
    seed = (seed + 0x6d2b79f5) | 0;
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

// ---------------------------------------------------------------- the pool

/**
 * A connection pool, and a database server that does not care about your
 * promises. `query` holds a connection for the whole of its duration;
 * aborting the caller does not shorten it. Only the statement timeout does,
 * and only when one was set.
 */
class Pool {
  constructor(size, metrics) {
    this.available = size;
    this.waiters = [];
    this.inUse = 0;
    this.m = metrics;
  }

  acquire() {
    if (this.available > 0) {
      this.available--;
      return Promise.resolve();
    }
    return new Promise((resolve) => this.waiters.push(resolve));
  }

  release() {
    const next = this.waiters.shift();
    if (next) next();
    else this.available++;
  }

  async query(durationMs, deadline, useStatementTimeout) {
    await this.acquire();
    try {
      // Checked out. Everything below happens inside the transaction, which
      // is the only place a statement timeout means anything.
      const now = Date.now();
      if (deadline !== null && deadline - now < SLACK_MS) {
        // The request that queued for this connection died while it was
        // queueing. Give the connection straight back rather than spend a
        // whole service time on a corpse. Under overload this is where most
        // of the recovered capacity comes from.
        this.m.abandoned++;
        return false;
      }

      this.inUse++;
      try {
        let stmt = null;
        if (deadline !== null && useStatementTimeout) {
          // Derived from the SAME number as the application budget. Two
          // independently chosen timeouts is how you get a service that
          // sheds load while the database stays pinned.
          stmt = Math.max(0, deadline - now - SLACK_MS);
        }
        if (stmt === null || stmt >= durationMs) {
          await sleep(durationMs);
          return true;
        }
        await sleep(stmt);
        this.m.killed++;
        return false;
      } finally {
        this.inUse--;
      }
    } finally {
      this.release();
    }
  }
}

// ------------------------------------------------------------- the metrics

function newMetrics() {
  return {
    ok: 0, timedOut: 0,
    zombie: 0, killed: 0, abandoned: 0,
    cLatencies: [], gauge: [],
  };
}

function percentile(sorted, p) {
  if (!sorted.length) return 0;
  const k = Math.round((p / 100) * (sorted.length - 1));
  return sorted[Math.min(k, sorted.length - 1)];
}

const mean = (v) => (v.length ? v.reduce((a, b) => a + b, 0) / v.length : 0);

// -------------------------------------------------------------- the hops

class Expired extends Error {}

/**
 * Wait for `promise`, but stop waiting when `signal` aborts. The promise
 * itself is NOT cancelled -- there is no such thing in JavaScript. This
 * function is the whole lesson in six lines: an abort ends your await, and
 * the work carries on without you.
 */
function untilAborted(promise, signal) {
  if (!signal) return promise;
  return Promise.race([
    promise,
    new Promise((_, reject) => {
      if (signal.aborted) return reject(new Expired('aborted'));
      signal.addEventListener('abort', () => reject(new Expired('aborted')), { once: true });
    }),
  ]);
}

async function serviceC(pool, m, slow, deadline, gatewayDeadline, signal, useStatementTimeout) {
  if (deadline !== null && deadline - Date.now() < SLACK_MS) {
    // Refuse to START work that cannot finish. A request rejected here costs
    // no pool slot, no queue position, nothing at all -- the cheapest win in
    // the entire layer.
    throw new Expired('no budget left at C');
  }

  await sleep(HOP_OVERHEAD_MS);

  const duration = slow ? C_SERVICE_SLOW_MS : C_SERVICE_FAST_MS;
  const started = Date.now();

  const running = pool.query(duration, deadline, useStatementTimeout).then((completed) => {
    const done = Date.now();
    m.cLatencies.push(done - started);
    if (completed && done > gatewayDeadline) m.zombie++;
    return completed;
  });

  // The forgotten-{signal} moment made explicit: we stop waiting, the query
  // does not stop running.
  await untilAborted(running, signal);
}

async function serviceB(pool, m, slow, deadline, gatewayDeadline, signal, useStatementTimeout) {
  if (deadline !== null && deadline - Date.now() < SLACK_MS) {
    throw new Expired('no budget left at B');
  }

  await sleep(HOP_OVERHEAD_MS);

  let outboundSignal;
  if (deadline !== null) {
    // budget_out = budget_in - elapsed_here - slack, expressed the way the
    // platform wants it. AbortSignal.any is the piece that makes this
    // compose: the callee stops for whichever reason comes first.
    const out = Math.max(0, deadline - Date.now() - SLACK_MS);
    outboundSignal = AbortSignal.any([signal, AbortSignal.timeout(out)]);
  } else {
    // The bug, and it reads as completely reasonable: a constant, the same
    // one the gateway used, chosen once and copied down the chain.
    outboundSignal = AbortSignal.timeout(GATEWAY_BUDGET_MS);
  }

  await serviceC(pool, m, slow, deadline, gatewayDeadline, outboundSignal, useStatementTimeout);
}

async function gateway(pool, m, slow, propagate, useStatementTimeout) {
  // The gateway's deadline is a per-request local. `gatewayDeadline` is
  // threaded through the chain purely so the measurement can tell a zombie
  // from a completion; `deadline` is the one the hops are allowed to read,
  // and in the naive variant it is deliberately null.
  const deadline = Date.now() + GATEWAY_BUDGET_MS;
  const signal = AbortSignal.timeout(GATEWAY_BUDGET_MS);
  try {
    await serviceB(pool, m, slow, propagate ? deadline : null, deadline, signal, useStatementTimeout);
    m.ok++;
  } catch (e) {
    // A TimeoutError from AbortSignal.timeout and our own Expired are the
    // same outcome from the gateway's seat, with very different costs.
    m.timedOut++;
  }
}

// -------------------------------------------------------------- the driver

async function runVariant(slowFraction, propagate, useStatementTimeout) {
  const m = newMetrics();
  const pool = new Pool(C_POOL_SIZE, m);
  const rng = mulberry32(20250502);

  const gaugeTimer = setInterval(() => m.gauge.push(pool.inUse), GAUGE_EVERY_MS);

  const begin = Date.now();
  const end = begin + DURATION_MS;
  let at = begin;
  const inFlight = [];

  for (;;) {
    at += (-Math.log(1 - rng()) / RATE) * 1000;
    if (at > end) break;
    const wait = at - Date.now();
    if (wait > 0) await sleep(wait);
    const slow = rng() < slowFraction;
    inFlight.push(gateway(pool, m, slow, propagate, useStatementTimeout));
  }

  await Promise.allSettled(inFlight);
  // Drain. Zombies are by definition still running after everyone gave up,
  // so a report taken at the end of the load would undercount them.
  await sleep(C_SERVICE_SLOW_MS + 300);
  clearInterval(gaugeTimer);
  return m;
}

function row(label, m) {
  const seconds = DURATION_MS / 1000;
  const total = m.ok + m.timedOut;
  const success = total ? (100 * m.ok) / total : 0;
  const p99 = percentile([...m.cLatencies].sort((a, b) => a - b), 99);
  return [
    label.padEnd(28),
    `${success.toFixed(1)}%`.padStart(10),
    (m.zombie / seconds).toFixed(1).padStart(9),
    `${mean(m.gauge).toFixed(1)}/${C_POOL_SIZE}`.padStart(14),
    p99.toFixed(0).padStart(9),
    (m.killed / seconds).toFixed(1).padStart(9),
    (m.abandoned / seconds).toFixed(1).padStart(11),
  ].join(' ');
}

const HEADER = [
  'variant'.padEnd(28), 'gw success'.padStart(10), 'zombie/s'.padStart(9),
  'C pool in use'.padStart(14), 'C p99 ms'.padStart(9), 'killed/s'.padStart(9),
  'gaveback/s'.padStart(11),
].join(' ');

async function main() {
  const fastDemand = (RATE * (1 - SLOW_FRACTION) * C_SERVICE_FAST_MS) / 1000;
  const slowDemand = (RATE * SLOW_FRACTION * C_SERVICE_SLOW_MS) / 1000;
  console.log(`Deadline propagation through gateway -> serviceB -> serviceC (node ${process.version}).`);
  console.log(`Gateway budget ${GATEWAY_BUDGET_MS}ms, slack ${SLACK_MS}ms/hop, C pool ${C_POOL_SIZE}, offered ${RATE} rps for ${DURATION_MS / 1000}s.`);
  console.log(`When C is unwell, ${SLOW_FRACTION * 100}% of queries take ${C_SERVICE_SLOW_MS}ms and the rest take ${C_SERVICE_FAST_MS}ms.`);
  console.log(`Demand on the pool is then ${slowDemand.toFixed(1)} + ${fastDemand.toFixed(1)} = ${(slowDemand + fastDemand).toFixed(1)} connection-seconds per second`);
  console.log(`against ${C_POOL_SIZE} available, i.e. rho = ${((slowDemand + fastDemand) / C_POOL_SIZE).toFixed(2)}. None of the slow queries can beat the budget.\n`);
  console.log(HEADER);
  console.log('-'.repeat(HEADER.length));

  const variants = [
    ['1 naive, C healthy', 0, false, false],
    ['2 naive, C slow', SLOW_FRACTION, false, false],
    ['3 propagated', SLOW_FRACTION, true, false],
    ['4 propagated + stmt timeout', SLOW_FRACTION, true, true],
  ];
  for (const [label, frac, prop, stmt] of variants) {
    const m = await runVariant(frac, prop, stmt);
    console.log(row(label, m));
  }

  console.log();
  console.log('Rows 2 and 3: AbortSignal.any composes the inbound signal with a');
  console.log('locally computed one, so B and C stop queueing work whose caller has');
  console.log('already gone, and hand connections back the moment they find one');
  console.log("checked out for a dead request ('gaveback/s').");
  console.log();
  console.log('Rows 3 and 4 are the part the platform cannot do for you. An');
  console.log('AbortSignal aborts an AWAIT. It does not reach into a database');
  console.log('server, a libuv thread-pool job, or anything else that already has');
  console.log('your work. Bounding the work where the work is -- row 4 -- is the');
  console.log('half you have to write yourself, in every runtime in this layer.');
}

main();
